docs/polaris/queueing-and-running-jobs/example-job-scripts.md

```diff
@@ -76,26 +111,33 @@ cd ${PBS_O_WORKDIR}
 # MPI and OpenMP settings
 NNODES=`wc -l < $PBS_NODEFILE`
 NRANKS_PER_NODE=$(nvidia-smi -L | wc -l)
-NDEPTH=8
+NDEPTH=16
 NTHREADS=1
 
 NTOTRANKS=$(( NNODES * NRANKS_PER_NODE ))
 echo "NUM_OF_NODES= ${NNODES} TOTAL_NUM_RANKS= ${NTOTRANKS} RANKS_PER_NODE= ${NRANKS_PER_NODE} THREADS_PER_RANK= ${NTHREADS}"
 
 # For applications that internally handle binding MPI/OpenMP processes to GPUs
-mpiexec -n ${NTOTRANKS} --ppn ${NRANKS_PER_NODE} --depth=${NDEPTH} --cpu-bind depth --env OMP_NUM_THREADS=${NTHREADS} -env OMP_PLACES=threads ./hello_affinity
+mpiexec -n ${NTOTRANKS} --ppn ${NRANKS_PER_NODE} --depth=${NDEPTH} --cpu-bind numa --env OMP_NUM_THREADS=${NTHREADS} -env OMP_PLACES=threads ./hello_affinity
 
 # For applications that need mpiexec to bind MPI ranks to GPUs
-#mpiexec -n ${NTOTRANKS} --ppn ${NRANKS_PER_NODE} --depth=${NDEPTH} --cpu-bind depth --env OMP_NUM_THREADS=${NTHREADS} -env OMP_PLACES=threads ./set_affinity_gpu_polaris.sh ./hello_affinity
+#mpiexec -n ${NTOTRANKS} --ppn ${NRANKS_PER_NODE} --depth=${NDEPTH} --cpu-bind numa --env OMP_NUM_THREADS=${NTHREADS} -env OMP_PLACES=threads ./set_affinity_gpu_polaris.sh ./hello_affinity
 ```
 
-The OpenMP-related options are not needed if your application does not use
-OpenMP. Nothing additional is required on the `mpiexec` command for
-applications that internally manage GPU devices and handle the binding of
-MPI/OpenMP processes to GPUs. A small helper script is available for those with
-applications that rely on MPI to handle the binding of MPI ranks to GPUs. Some
-notes on this helper script and other key differences with the early CPU
-example follow.
+As in the previous example, the MPI ranks are spaced apart assuming the user wants to
+utilize all 64 logical cores (achieved by setting `NTHREADS=$NDEPTH` and `--cpu-bind numa`
+here). Set `NDEPTH=8; NTHREADS=8` and use `--cpu-bind depth` or `core` to spawn only 1
```

Comment:
If I understand correctly, `--cpu-bind core` refers to binding each MPI rank to a single core. I cannot understand how it works with `NDEPTH=8; NTHREADS=8`.

Probably using two code sections makes it more clear both cases w/ and w/o hardware threads. Words are just confusing.

Reply:
`--cpu-bind core `
```
NUM_OF_NODES= 1 TOTAL_NUM_RANKS= 4 RANKS_PER_NODE= 4 THREADS_PER_RANK= 8
“RANK= 3 LOCAL_RANK= 3 gpu= 0”
“RANK= 1 LOCAL_RANK= 1 gpu= 2”
“RANK= 2 LOCAL_RANK= 2 gpu= 1”
“RANK= 0 LOCAL_RANK= 0 gpu= 3”
rnk= 0 :  # of devices detected= 1
To affinity and beyond!! nname= x3005c0s13b1n0  rnk= 0  list_cores= (0-7,32-39)  num_devices= 1

To affinity and beyond!! nname= x3005c0s13b1n0  rnk= 1  list_cores= (8-15,40-47)  num_devices= 1

To affinity and beyond!! nname= x3005c0s13b1n0  rnk= 2  list_cores= (16-23,48-55)  num_devices= 1

To affinity and beyond!! nname= x3005c0s13b1n0  rnk= 3  list_cores= (24-31,56-63)  num_devices= 1

```

`--cpu-bind depth `
```
NUM_OF_NODES= 1 TOTAL_NUM_RANKS= 4 RANKS_PER_NODE= 4 THREADS_PER_RANK= 8
“RANK= 2 LOCAL_RANK= 2 gpu= 1”
“RANK= 3 LOCAL_RANK= 3 gpu= 0”
“RANK= 0 LOCAL_RANK= 0 gpu= 3”
“RANK= 1 LOCAL_RANK= 1 gpu= 2”
rnk= 0 :  # of devices detected= 1
To affinity and beyond!! nname= x3005c0s13b0n0  rnk= 0  list_cores= (0-7)  num_devices= 1

To affinity and beyond!! nname= x3005c0s13b0n0  rnk= 1  list_cores= (8-15)  num_devices= 1

To affinity and beyond!! nname= x3005c0s13b0n0  rnk= 2  list_cores= (16-23)  num_devices= 1

To affinity and beyond!! nname= x3005c0s13b0n0  rnk= 3  list_cores= (24-31)  num_devices= 1

```

Reply:
`-d 8 --cpu-bind core` gives you 8 physical cores per MPI rank

Reply:
Crazy mpich. The documentation clearly says
```
           --cpu-bind
               Bind ranks to CPUs

               Keywords,

               * core - Bind each process to a core.
               * depth - Bind ranks to the number of threads in the 'depth' argument.
               * none - No CPU binding.
               * numa, socket, core, thread - Bind ranks to the specified hardware.
               * list - Bind ranks to colon-separated range lists of CPUs.
               * mask - Bind ranks to comma-separated bitmasks of CPUs.
```

Reply:
depends on your definition of "core" I suppose. But yeah, it doesnt say how the behavior of `--depth` changes with several of those binding options

Reply:
I believe `--depth` is only relevant when using `-cpu-bind depth`.
Could you add your example instead of modifying the existing one.

The old example seems to be MPI + GPU without OpenMP threading
Your example sees to be MPI + GPU with OpenMP threading on hardware threads.
it will be good to have an example of MPI + GPU with OpenMP threading wihtout hardware threads.
Blending three example in one is very hard for user to follow.

Reply:
The old example includes OpenMP threading too, does it not? That's what's shown on the published LCF docs page.

Reply:
> I believe `--depth` is only relevant when using `--cpu-bind depth`

You would think so, but that is wrong. It affects `core` and `numa` binding, but not `socket`, `none`, nor `thread`. I didnt test `mask` or `list`

Reply:
@zippylab The old example has `NTHREADS=1` set.

Reply:
But the old MPI-GPU example had `NDEPTH=8`, and `--env OMP_NUM_THREADS=${NTHREADS} -env OMP_PLACES=threads`. Which is all overkill if you arent using threads at all, so you are suggesting that the old example cut all that out? No need to specify depth if you dont plan on threading; easier to understand to use `--cpu-bind=numa` without `-d`

Reply:
32 cores sit on 4 numa. So using NDEPTH=8 per MPI rank is to prevent clustering all the MPI processes in a single numa. `--env OMP_NUM_THREADS=${NTHREADS} -env OMP_PLACES=threads` can play positively or negatively depending on an app.
`--depth 8 --cpu-bind depth` and `--cpu-bind numa` probably achieves the same goal. Only a matter of preference.